File: rust/src/bin/1218_longest_arithmetic_subsequence_of_given_difference.rs
fn main() {
    assert_eq!(longest_subsequence(vec![3, 4, -3, -2, -4], -5), 2);
    assert_eq!(longest_subsequence(vec![1, 2, 3, 4], 1), 4);
    assert_eq!(longest_subsequence(vec![1, 3, 5, 7], 1), 1);
    assert_eq!(longest_subsequence(vec![1, 5, 7, 8, 5, 3, 4, 2, 1], -2), 4);
}

/*
 * [1,3,5,7]
 *
 * O(n), O(n)
 */
pub fn longest_subsequence(arr: Vec<i32>, difference: i32) -> i32 {
    type Map = std::collections::HashMap<i32, i32>;
    let mut dp = Map::new();

    let mut lst = 0;
    for i in 0..arr.len() {
        let nr = *dp.get(&(arr[i] - difference)).unwrap_or(&0) + 1;

        dp.insert(arr[i], nr);

        lst = lst.max(nr);
    }

    lst
}

File: ts/src/ts/1218_longest_arithmetic_subsequence_of_given_difference.ts
import assertEqual from './assert';

/*
 *
 *
 * O(n), O(n)
 */
function longestSubsequence(arr: number[], difference: number): number {
  const cache = new Map<number, number>();

  let longest = 0;

  for (let x of arr) {
    let nr = 1 + (cache.get(x - difference) || 0);

    cache.set(x, nr);

    longest = Math.max(longest, nr);
  }

  return longest;
};

assertEqual(longestSubsequence([1, 2, 3, 4], 1), 4)
assertEqual(longestSubsequence([1, 3, 5, 7], 1), 1)
assertEqual(longestSubsequence([1, 5, 7, 8, 5, 3, 4, 2, 1], -2), 4)
assertEqual(longestSubsequence([3, 4, -3, -2, -4], -5), 2)




